kbncontent.go

```diff
@@ -0,0 +1,319 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+// Package kbncontent implements routines for analyzing Kibana content.
+//
+// It provides information about Kibana assets and a single source of truth for what things are legacy and should no longer be used.
+package kbncontent
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/stretchr/objx"
+)
+
+// Describes a visualization.
+// In this case, "visualization" means anything which can be embedded in a dashboard.
+type VisualizationDescriptor struct {
+	Doc    map[string]interface{}
+	SoType string
+	Link   string
+}
+
+func (v VisualizationDescriptor) tryDocumentPaths(paths []string) string {
+	m := objx.Map(v.Doc)
+
+	for _, path := range paths {
+		if m.Get(path).IsStr() {
+			return m.Get(path).Str()
+		}
+	}
+
+	return ""
+}
+
+// root-level visualization type
+// currently empty for Lens
+func (v VisualizationDescriptor) Type() string {
+	if v.SoType != "visualization" {
+		return ""
+	}
+
+	return v.tryDocumentPaths([]string{
+		"attributes.type",
+		"attributes.visState.type",
+		"embeddableConfig.savedVis.type", // by-value dashboard panel
+	})
+}
+
+// name of the visualization editor
+func (v VisualizationDescriptor) Editor() string {
+	if v.SoType == "lens" {
+		return "Lens"
+	}
+
+	if v.SoType == "map" {
+		return "Maps"
+	}
+
+	if v.SoType == "search" {
+		return "Discover"
+	}
+
+	if v.SoType == "visualization" {
+		if v.Type() == "metrics" {
+			return "TSVB"
+		}
+
+		if v.Type() == "vega" {
+			return "Vega"
+		}
+
+		if v.Type() == "timelion" {
+			return "Timelion"
+		}
+
+		return "Aggs-based"
+	}
+
+	return "Unknown"
```

Comment:
Nit. Consider using `switch`/`case` here.
```suggestion
	switch v.SoType {
	case "lens":
		return "Lens"
	case "map":
		return "Maps"
	case "search":
		return "Discover"
	case "visualization":
		switch v.Type() {
		case "metrics":
			return "TSVB"
		case "vega":
			return "Vega"
		case "timelion":
			return "Timelion"
		default:
			return "Aggs-based"
		}
	}

	return "Unknown"
```